internal/app/app.go

```diff
@@ -0,0 +1,153 @@
+// Package app exports a D-Bus object to apply proxy settings.
+package app
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/godbus/dbus/v5"
+	"github.com/godbus/dbus/v5/introspect"
+	log "github.com/sirupsen/logrus"
+	"github.com/ubuntu/decorate"
+	"github.com/ubuntu/ubuntu-proxy-manager/internal/proxy"
+)
+
+const dbusObjectPath = "/com/ubuntu/ProxyManager"
+const dbusInterface = "com.ubuntu.ProxyManager"
+
+// DryRun is a context key to indicate that we should not apply the proxy
+// settings. It is intended to be used in tests only.
+type DryRun struct{}
+
+type busObject struct {
+	ctx    context.Context
+	cancel context.CancelCauseFunc
+
+	mu sync.Mutex
+}
+
+type options struct {
+	timeout time.Duration
+	initCtx context.Context
+}
+type option func(*options)
+
+// App is the main application object.
+type App struct {
+	conn    *dbus.Conn
+	obj     *busObject
+	timeout time.Duration
+}
+
+// Apply is a function called via D-Bus to apply the system proxy settings.
+func (b *busObject) Apply(http, https, ftp, socks, no, mode string) *dbus.Error {
+	// Methods calls spin up separate goroutines, so ensure we don't run them in parallel
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	log.Infof("Called Apply(%s, %s, %s, %s, %s, %s)", http, https, ftp, socks, no, mode)
+
+	p, err := proxy.New(b.ctx, http, https, ftp, socks, no, mode)
+	// We need to cancel the context in a deferred function to get the final
+	// state of the error variable.
+	defer func() { b.cancel(err) }()
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	if b.ctx.Value(DryRun{}) == true {
```

Comment:
I would rather put the proxy in "dry run" mode in New() so that you can test it later on the line (with/without dry run), but get the same output (like sub files on disk) for integration and proxy tests.

Reply:
I did use `afero` in the past and kind of regretted it. It’s moving too much away from stdlib, which has now a standard fs interface if you want for mocking it (but TBH, I was never brave enough/got enough time) to really try it. I wonder how it will work with our golden files approach though. I guess you can experiment :p

My initial comment basically was to move the dryrun skip in the apply() call for the proxy itself.